BumpBox<FnOnce> cannot be called

Compiled with cargo 1.88.0-nightly (0e93c5bf6 2025-04-05)
```rust
use bump_scope::{Bump, BumpBox};
fn main() {
    let arena: Bump = Bump::new();

    // Compile error
    let x: BumpBox<dyn FnOnce()> = arena.alloc(|| {});
    (x)();

    // Works fine
    let x: Box<dyn FnOnce()> = Box::new(|| {});
    (x)();
}
```

# Error

```
error[E0161]: cannot move a value of type `dyn FnOnce()`
 --> src/bin/bump_scope.rs:7:5
  |
7 |     (x)();
  |     ^^^ the size of `dyn FnOnce()` cannot be statically determined

error[E0507]: cannot move out of dereference of `BumpBox<'_, dyn FnOnce()>`
 --> src/bin/bump_scope.rs:7:5
  |
7 |     (x)();
  |     ^^^--
  |     |
  |     value moved due to this call
  |     move occurs because value has type `dyn FnOnce()`, which does not implement the `Copy` trait
  |
note: this value implements `FnOnce`, which causes it to be moved when called
 --> src/bin/bump_scope.rs:7:5
  |
7 |     (x)();
  |     ^^^

Some errors have detailed explanations: E0161, E0507.
For more information about an error, try `rustc --explain E0161`.
error: could not compile `sy41` (bin "bump_scope") due to 2 previous errors

```

--- a/.vscode/cspell.json
+++ b/.vscode/cspell.json
@@ -56,6 +56,7 @@
     "haha",
     "hashbrown",
     "Hasher",
+    "impls",
     "inplace",
     "insertable",
     "justfile",

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,8 @@
 # Changelog
 
+## Unreleased
+- **added:** `BumpBox<T>` now implements the `Fn*` traits if `T` does (just like `Box`). This requires the new `"nightly-fn-traits"` feature.
+
 ## 0.16.3 (2025-02-21)
 - **added:** `capacity` field to the `Debug` representation of `Stats` and `Bump(Scope(Guard(Root)))`
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -56,6 +56,9 @@ nightly-exact-size-is-empty = []
 ## Implements `TrustedLen` for some iterators.
 nightly-trusted-len = []
 
+## Implements `Fn*` traits for `BumpBox<T>`. Makes `BumpBox<T: FnOnce>` callable. Requires alloc crate.
+nightly-fn-traits = []
+
 [dev-dependencies]
 trybuild = "1.0.90"
 expect-test = "1.4.1"

--- a/README.md
+++ b/README.md
@@ -191,6 +191,7 @@ assert_eq!(bump.stats().allocated(), 4);
   You can unsize a `BumpBox` in stable without this feature using [`unsize_bump_box`].
 * **`nightly-exact-size-is-empty`** —  Implements `is_empty` manually for some iterators.
 * **`nightly-trusted-len`** —  Implements `TrustedLen` for some iterators.
+* **`nightly-fn-traits`** —  Implements `Fn*` traits for `BumpBox<T>`. Makes `BumpBox<T: FnOnce>` callable. Requires alloc crate.
 
 ## Bumping upwards or downwards?
 Bump direction is controlled by the generic parameter `const UP: bool`. By default, `UP` is `true`, so the allocator bumps upwards.

--- a/src/bump_box.rs
+++ b/src/bump_box.rs
@@ -13,6 +13,9 @@ use core::{
     str,
 };
 
+#[cfg(feature = "nightly-fn-traits")]
+use core::marker::Tuple;
+
 #[cfg(feature = "std")]
 use alloc::{string::String, vec::Vec};
 
@@ -3088,6 +3091,57 @@ impl<T: ?Sized + std::io::BufRead> std::io::BufRead for BumpBox<'_, T> {
     #[inline(always)]
     fn read_line(&mut self, buf: &mut String) -> std::io::Result<usize> {
         T::read_line(self, buf)
+    }
+}
+
+#[cfg(feature = "nightly-fn-traits")]
+impl<Args: Tuple, F: FnOnce<Args> + ?Sized> FnOnce<Args> for BumpBox<'_, F> {
+    type Output = <F as FnOnce<Args>>::Output;
+
+    extern "rust-call" fn call_once(self, args: Args) -> Self::Output {
+        use alloc::boxed::Box;
+
+        use allocator_api2::alloc::{AllocError, Allocator};
+
+        struct NoopAllocator;
+
+        // SAFETY:
+        // An allocator that always fails allocation and does nothing on deallocation
+        // satisfies the safety invariants.
+        unsafe impl Allocator for NoopAllocator {
+            fn allocate(&self, _layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+                Err(AllocError)
+            }
+
+            unsafe fn deallocate(&self, _ptr: NonNull<u8>, _layout: Layout) {}
+        }
+
+        let ptr = self.into_raw().as_ptr();
+
+        // SAFETY:
+        // The allocator will only be used to call `deallocate` which does nothing
+        // just like the `BumpBox` would do no deallocation on drop.
+        //
+        // The `Box` or its allocator do not encode the lifetime of the `BumpBox` but
+        // it's fine here because it only lives in this function.
+        unsafe {
+            let boxed = Box::from_raw_in(ptr, NoopAllocator);
+            <Box<F, NoopAllocator> as FnOnce<Args>>::call_once(boxed, args)
+        }
+    }
+}
+
+#[cfg(feature = "nightly-fn-traits")]
+impl<Args: Tuple, F: FnMut<Args> + ?Sized> FnMut<Args> for BumpBox<'_, F> {
+    extern "rust-call" fn call_mut(&mut self, args: Args) -> Self::Output {
+        <F as FnMut<Args>>::call_mut(self, args)
+    }
+}
+
+#[cfg(feature = "nightly-fn-traits")]
+impl<Args: Tuple, F: Fn<Args> + ?Sized> Fn<Args> for BumpBox<'_, F> {
+    extern "rust-call" fn call(&self, args: Args) -> Self::Output {
+        <F as Fn<Args>>::call(self, args)
     }
 }
 

--- a/src/bump_box.rs
+++ b/src/bump_box.rs
@@ -3099,9 +3099,10 @@ impl<Args: Tuple, F: FnOnce<Args> + ?Sized> FnOnce<Args> for BumpBox<'_, F> {
     type Output = <F as FnOnce<Args>>::Output;
 
     extern "rust-call" fn call_once(self, args: Args) -> Self::Output {
-        use alloc::boxed::Box;
-
-        use allocator_api2::alloc::{AllocError, Allocator};
+        use alloc::{
+            alloc::{AllocError, Allocator},
+            boxed::Box,
+        };
 
         struct NoopAllocator;
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -5,10 +5,15 @@
 // Especially `BumpBox` methods, vectors, strings, `polyfill` and `tests/from_std` are based on code from the standard library.
 
 #![no_std]
-#![cfg_attr(feature = "nightly-allocator-api", feature(allocator_api, vec_into_raw_parts))]
+#![cfg_attr(
+    any(feature = "nightly-allocator-api", feature = "nightly-fn-traits"),
+    feature(allocator_api)
+)]
+#![cfg_attr(feature = "nightly-allocator-api", feature(vec_into_raw_parts))]
 #![cfg_attr(feature = "nightly-coerce-unsized", feature(coerce_unsized, unsize))]
 #![cfg_attr(feature = "nightly-exact-size-is-empty", feature(exact_size_is_empty))]
 #![cfg_attr(feature = "nightly-trusted-len", feature(trusted_len))]
+#![cfg_attr(feature = "nightly-fn-traits", feature(fn_traits, tuple_trait, unboxed_closures))]
 #![cfg_attr(
     test,
     feature(
@@ -242,6 +247,7 @@
 //!   You can unsize a `BumpBox` in stable without this feature using [`unsize_bump_box`].
 //! * **`nightly-exact-size-is-empty`** —  Implements `is_empty` manually for some iterators.
 //! * **`nightly-trusted-len`** —  Implements `TrustedLen` for some iterators.
+//! * **`nightly-fn-traits`** —  Implements `Fn*` traits for `BumpBox<T>`. Makes `BumpBox<T: FnOnce>` callable. Requires alloc crate.
 //!
 //! # Bumping upwards or downwards?
 //! Bump direction is controlled by the generic parameter `const UP: bool`. By default, `UP` is `true`, so the allocator bumps upwards.
@@ -275,7 +281,7 @@
 #[cfg(any(feature = "std", test))]
 extern crate std;
 
-#[cfg(feature = "alloc")]
+#[cfg(any(feature = "alloc", feature = "nightly-fn-traits"))]
 extern crate alloc;
 
 mod allocator;

--- a/src/tests/fn_traits.rs
+++ b/src/tests/fn_traits.rs
@@ -1,0 +1,55 @@
+use std::{panic::catch_unwind, string::String};
+
+use crate::{Bump, BumpBox};
+
+#[test]
+fn fn_once() {
+    let bump: Bump = Bump::new();
+
+    let x: BumpBox<dyn FnOnce()> = bump.alloc(|| {});
+    (x)();
+
+    // making sure the string is dropped
+    let string = String::from("hello world");
+    let x: BumpBox<dyn FnOnce()> = bump.alloc(move || drop(string));
+    (x)();
+
+    let string = String::from("hello world");
+    let x: BumpBox<dyn FnOnce()> = bump.alloc(move || drop(string));
+    drop(x);
+
+    #[allow(unused_variables)]
+    catch_unwind(|| {
+        let string = String::from("hello world");
+        let x: BumpBox<dyn FnOnce()> = bump.alloc(move || {
+            panic!("a");
+            #[allow(unreachable_code)]
+            drop(string);
+        });
+        (x)();
+    })
+    .unwrap_err();
+}
+
+#[test]
+fn does_impl() {
+    fn impls_fn_once<T: FnOnce()>(_: &T) {}
+    fn impls_fn_mut<T: FnMut()>(_: &T) {}
+    fn impls_fn<T: Fn()>(_: &T) {}
+
+    let bump: Bump = Bump::new();
+
+    let f = bump.alloc(|| {});
+    impls_fn_once(&f);
+    impls_fn_mut(&f);
+    impls_fn(&f);
+
+    let mut string = String::from("hello");
+    let f = bump.alloc(|| string.push('x'));
+    impls_fn_once(&f);
+    impls_fn_mut(&f);
+
+    let string = String::from("hello");
+    let f = bump.alloc(|| drop(string));
+    impls_fn_once(&f);
+}

--- a/src/tests/fn_traits.rs
+++ b/src/tests/fn_traits.rs
@@ -52,4 +52,18 @@ fn does_impl() {
     let string = String::from("hello");
     let f = bump.alloc(|| drop(string));
     impls_fn_once(&f);
+
+    let f: BumpBox<dyn Fn()> = bump.alloc(|| {});
+    impls_fn_once(&f);
+    impls_fn_mut(&f);
+    impls_fn(&f);
+
+    let mut string = String::from("hello");
+    let f: BumpBox<dyn FnMut()> = bump.alloc(|| string.push('x'));
+    impls_fn_once(&f);
+    impls_fn_mut(&f);
+
+    let string = String::from("hello");
+    let f: BumpBox<dyn FnOnce()> = bump.alloc(|| drop(string));
+    impls_fn_once(&f);
 }

--- a/src/tests/mod.rs
+++ b/src/tests/mod.rs
@@ -36,7 +36,7 @@ mod chunk_size;
 #[cfg(feature = "nightly-coerce-unsized")]
 mod coerce_unsized;
 mod fixed_bump_vec;
-#[cfg(feature = "nightly-fn-traits")]
+#[cfg(all(feature = "nightly-fn-traits", feature = "nightly-coerce-unsized"))]
 mod fn_traits;
 mod from_std;
 mod grow_vec;

--- a/src/tests/mod.rs
+++ b/src/tests/mod.rs
@@ -36,6 +36,8 @@ mod chunk_size;
 #[cfg(feature = "nightly-coerce-unsized")]
 mod coerce_unsized;
 mod fixed_bump_vec;
+#[cfg(feature = "nightly-fn-traits")]
+mod fn_traits;
 mod from_std;
 mod grow_vec;
 mod into_flattened;

